// neo4j/lib/Node.js
/*** Generated by streamline 0.4.0 (callbacks) - DO NOT EDIT ***/
var __rt=require('streamline/lib/callbacks/runtime').runtime(__filename),__func=__rt.__func,__cb=__rt.__cb,__tryCatch=__rt.__tryCatch,__apply=__rt.__apply;
/*     1 */ (function() {
/*     2 */   var Node, Path, PropertyContainer, Relationship, adjustError, flows, status, util, __hasProp = {
/*     3 */   }.hasOwnProperty, __extends = function(child, parent) {
/*     4 */     for (var key in parent) {
/*     4 */       if (__hasProp.call(parent, key)) {
/*     4 */         child[key] = parent[key];
                  };
                };
/*     4 */     function ctor() {
/*     4 */       this.constructor = child;
                };
/*     4 */     ctor.prototype = parent.prototype;
/*     4 */     child.prototype = new ctor();
/*     4 */     child.__super__ = parent.prototype;
/*     4 */     return child;
              };
/*     6 */   flows = require("streamline/lib/util/flows");
/*     8 */   status = require("http-status");
/*    10 */   util = require("./util");
/*    12 */   adjustError = util.adjustError;
/*    14 */   PropertyContainer = require("./PropertyContainer");
/*    16 */   Relationship = require("./Relationship");
/*    18 */   Path = require("./Path");
/*    20 */   module.exports = Node = (function(_super) {
/*    22 */     __extends(Node, _super);
/*    24 */     function Node(db, data) {
/*    25 */       Node.__super__.constructor.call(this, db, data);
                };
/*    28 */     Node.prototype.toString = function() {
/*    29 */       if (this.exists) {
/*    30 */         return ("node @" + this.id);
                  }
/*    31 */        else {
/*    32 */         return (("unsaved node (" + (JSON.stringify(this.data, null, 4))) + ")");
                  }
                ;
                };
/*    36 */     Node.prototype.save = function Node_prototype_save__1(_) {
                  var response, services, __this = this;
                  var __frame = {
                    name: "Node_prototype_save__1",
                    line: 36
                  };
                  return __func(_, this, arguments, Node_prototype_save__1, 0, __frame, function __$Node_prototype_save__1() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_save__1() {
                          return (function __$Node_prototype_save__1(__then) {
/*    39 */                 if (__this.exists) {
/*    40 */                   return __this._request.put({
/*    41 */                     uri: (("" + __this.self) + "/properties"),
/*    42 */                     json: __this.data
                              }, __cb(_, __frame, 4, 21, function ___(__0, __1) {
/*    40 */                     response = __1;
/*    44 */                     if ((response.statusCode !== status.NO_CONTENT)) {
/*    45 */                       switch (response.statusCode) {
/*    46 */                       case status.BAD_REQUEST:
/*    46 */                         return _(new Error("Invalid data sent"));
/*    47 */                       case status.NOT_FOUND:
/*    49 */                         return _(new Error("Node not found"));
/*    50 */                         default:
/*    52 */                         return _(response);
/*    53 */                       };
                                }
                              ;
                                __then();
                              }, true));
                            }
                             else {
/*    57 */                   return __this.db.getServices(__cb(_, __frame, 21, 21, function ___(__0, __2) {
/*    57 */                     services = __2;
/*    58 */                     return __this._request.post({
/*    59 */                       uri: services.node,
/*    60 */                       json: __this.data
                                }, __cb(_, __frame, 22, 21, function ___(__0, __3) {
/*    58 */                       response = __3;
/*    62 */                       if ((response.statusCode !== status.CREATED)) {
/*    63 */                         switch (response.statusCode) {
/*    64 */                         case status.BAD_REQUEST:
/*    64 */                           return _(new Error("Invalid data sent"));
/*    65 */                           default:
/*    67 */                           return _(response);
/*    68 */                         };
                                  }
                                ;
/*    71 */                       __this._data = response.body;
                                  __then();
                                }, true));
                              }, true));
                            }
                          ;
                          })(function __$Node_prototype_save__1() {
                            return _(null, __this);
                          });
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_save__1() {
                          if (error) {
/*    75 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*    79 */     Node.prototype["delete"] = function Node_prototype_delete__2(_, force) {
                  var relationship, relationships, __this = this, __arguments = arguments;
                  var __frame = {
                    name: "Node_prototype_delete__2",
                    line: 79
                  };
                  return __func(_, this, arguments, Node_prototype_delete__2, 0, __frame, function __$Node_prototype_delete__2() {
/*    81 */         if ((force == null)) {
/*    82 */           force = false;
                    }
                  ;
/*    84 */         if (!__this.exists) {
                      return _(null);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_delete__2() {
                          return (function __$Node_prototype_delete__2(__then) {
/*    88 */                 if (force) {
/*    89 */                   return __this.all(null, __cb(_, __frame, 10, 26, function ___(__0, __2) {
/*    89 */                     relationships = __2;
/*    98 */                     return (function __1(_) {
                                  var _i, _len, _results;
/*    92 */                       _results = [];
/*    93 */                       for (_i = 0, _len = relationships.length; (_i < _len); _i++) {
/*    94 */                         relationship = relationships[_i];
/*    95 */                         _results.push(relationship["delete"]());
                                  };
/*    97 */                       return _(null, _results);
/*    98 */                     })(__cb(_, __frame, 19, 27, function ___(__0, __3) {
/*    90 */                       return flows.collect(__cb(_, __frame, 11, 10, __then, true), __3);
                                }, true));
                              }, true));
                            }
                             else {
                              __then();
                            }
                          ;
                          })(__then);
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_delete__2() {
                          if (error) {
/*   101 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, function __$Node_prototype_delete__2() {
/*   103 */             return __apply(__cb(_, __frame, 24, 13, _, true), Node.__super__["delete"], __this, __arguments, 0);
                      });
                    });
                  });
                };
/*   106 */     Node.prototype.index = function Node_prototype_index__3(index, key, value, _) {
                  var encodedKey, encodedValue, response, services, url, version, __this = this;
                  var __frame = {
                    name: "Node_prototype_index__3",
                    line: 106
                  };
                  return __func(_, this, arguments, Node_prototype_index__3, 3, __frame, function __$Node_prototype_index__3() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_index__3() {
/*   109 */               if (!__this.exists) {
/*   110 */                 return _(new Error("Node must exists before indexing properties"));
                          }
                        ;
/*   112 */               return __this.db.getServices(__cb(_, __frame, 6, 19, function ___(__0, __1) {
/*   112 */                 services = __1;
/*   113 */                 return __this.db.getVersion(__cb(_, __frame, 7, 18, function ___(__0, __2) {
/*   113 */                   version = __2;
                              return (function __$Node_prototype_index__3(__then) {
/*   114 */                     if ((version <= 1.4)) {
/*   115 */                       encodedKey = encodeURIComponent(key);
/*   116 */                       encodedValue = encodeURIComponent(value);
/*   117 */                       url = ((((((("" + services.node_index) + "/") + index) + "/") + encodedKey) + "/") + encodedValue);
/*   118 */                       return __this._request.post({
/*   119 */                         url: url,
/*   120 */                         json: __this.self
                                  }, __cb(_, __frame, 12, 21, function ___(__0, __3) {
/*   118 */                         response = __3;
                                    __then();
                                  }, true));
                                }
                                 else {
/*   123 */                       return __this._request.post({
/*   124 */                         url: ((("" + services.node_index) + "/") + index),
/*   125 */                         json: {
/*   126 */                           key: key,
/*   127 */                           value: value,
/*   128 */                           uri: __this.self
                                    }
                                  }, __cb(_, __frame, 17, 21, function ___(__0, __4) {
/*   123 */                         response = __4;
                                    __then();
                                  }, true));
                                }
                              ;
                              })(function __$Node_prototype_index__3() {
/*   132 */                     if ((response.statusCode !== status.CREATED)) {
/*   133 */                       return _(response);
                                }
                              ;
                                __then();
                              });
                            }, true));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_index__3() {
                          if (error) {
/*   136 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   140 */     Node.prototype.createRelationshipTo = function Node_prototype_createRelationshipTo__4(otherNode, type, data, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_createRelationshipTo__4",
                    line: 140
                  };
                  return __func(_, this, arguments, Node_prototype_createRelationshipTo__4, 3, __frame, function __$Node_prototype_createRelationshipTo__4() {
/*   141 */         return __this._createRelationship(__this, otherNode, type, data, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   144 */     Node.prototype.createRelationshipFrom = function Node_prototype_createRelationshipFrom__5(otherNode, type, data, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_createRelationshipFrom__5",
                    line: 144
                  };
                  return __func(_, this, arguments, Node_prototype_createRelationshipFrom__5, 3, __frame, function __$Node_prototype_createRelationshipFrom__5() {
/*   145 */         return __this._createRelationship(otherNode, __this, type, data, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   148 */     Node.prototype._createRelationship = function Node_prototype__createRelationship__6(from, to, type, data, _) {
                  var createRelationshipURL, exception, message, otherNodeURL, response, _ref, __this = this;
                  var __frame = {
                    name: "Node_prototype__createRelationship__6",
                    line: 148
                  };
                  return __func(_, this, arguments, Node_prototype__createRelationship__6, 4, __frame, function __$Node_prototype__createRelationship__6() {
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype__createRelationship__6() {
/*   151 */               createRelationshipURL = from._data["create_relationship"];
/*   152 */               otherNodeURL = to.self;
                          return (function __$Node_prototype__createRelationship__6(__then) {
/*   153 */                 if ((((createRelationshipURL != null)) && otherNodeURL)) {
/*   154 */                   return __this._request.post({
/*   155 */                     url: createRelationshipURL,
/*   156 */                     json: {
/*   157 */                       to: otherNodeURL,
/*   158 */                       data: data,
/*   159 */                       type: type
                                }
                              }, __cb(_, __frame, 6, 21, function ___(__0, __3) {
/*   154 */                     response = __3;
                                return (function __$Node_prototype__createRelationship__6(__then) {
/*   162 */                       if ((response.statusCode !== status.CREATED)) {
/*   163 */                         _ref = (response.body || {
/*   163 */                         }), message = _ref.message, exception = _ref.exception;
                                    return (function __$Node_prototype__createRelationship__6(_) {
/*   164 */                           var __2 = message;
                                      if (__2) {
                                        return _(null, __2);
                                      }
                                    ;
                                      return (function __$Node_prototype__createRelationship__6(_) {
/*   164 */                             var __3 = exception;
                                        if (__3) {
                                          return _(null, __3);
                                        }
                                      ;
/*   173 */                             return (function __1(_) {
/*   165 */                               switch (response.statusCode) {
/*   166 */                               case status.BAD_REQUEST:
/*   166 */                                 return _(null, ((((((("Invalid createRelationship: " + from.id) + " ") + type) + " ") + to.id) + " w/ data: ") + (JSON.stringify(data))));
/*   167 */                               case status.CONFLICT:
/*   168 */                                 return _(null, "\"to\" node, or the node specified by the URI not found");
/*   169 */                                 default:
/*   170 */                                 return _(response);
/*   171 */                               };
                                          _();
/*   173 */                             })(__cb(_, __frame, 25, 47, _, true));
                                      })(__cb(_, __frame, -147, 7, function ___(__0, __5) {
/*   164 */                             var __4 = (message = __5);
                                        return _(null, __4);
                                      }, true));
                                    })(__cb(_, __frame, -147, 7, function __$Node_prototype__createRelationship__6() {
/*   174 */                           return _(new Error(message));
                                    }, true));
                                  }
                                   else {
                                    __then();
                                  }
                                ;
                                })(function __$Node_prototype__createRelationship__6() {
/*   176 */                       return _(null, new Relationship(__this.db, response.body, from, to));
                                });
                              }, true));
                            }
                             else {
/*   178 */                   return _(new Error("Failed to create relationship"));
                            }
                          ;
                          })(__then);
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype__createRelationship__6() {
                          if (error) {
/*   181 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   185 */     Node.prototype._getRelationships = function Node_prototype__getRelationships__7(direction, type, _) {
                  var prefix, relationshipsURL, resp, types, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype__getRelationships__7",
                    line: 185
                  };
                  return __func(_, this, arguments, Node_prototype__getRelationships__7, 2, __frame, function __$Node_prototype__getRelationships__7() {
                    _this = __this;
/*   188 */         types = null;
/*   189 */         if ((type != null)) {
/*   190 */           types = ((type instanceof Array) ? type : [type,]);
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype__getRelationships__7() {
/*   193 */               if ((types != null)) {
/*   194 */                 prefix = __this._data[(("" + direction) + "_typed_relationships")];
/*   195 */                 relationshipsURL = ((prefix != null) ? prefix.replace("{-list|&|types}", types.join("&")) : void 0);
                          }
/*   196 */                else {
/*   197 */                 relationshipsURL = __this._data[(("" + direction) + "_relationships")];
                          }
                        ;
/*   199 */               if (!relationshipsURL) {
/*   200 */                 return _(new Error("Couldn't find URL of relationships endpoint."));
                          }
                        ;
/*   202 */               return __this._request.get(relationshipsURL, __cb(_, __frame, 17, 15, function ___(__0, __1) {
/*   202 */                 resp = __1;
/*   203 */                 if ((resp.statusCode === status.NOT_FOUND)) {
/*   204 */                   return _(new Error("Node not found."));
                            }
                          ;
/*   206 */                 if ((resp.statusCode !== status.OK)) {
/*   207 */                   return _(new Error(("Unrecognized response code: " + resp.statusCode)));
                            }
                          ;
/*   209 */                 return _(null, resp.body.map(function(data) {
/*   210 */                   if ((_this.self === data.start)) {
/*   211 */                     return new Relationship(_this.db, data, _this, null);
                              }
/*   212 */                    else {
/*   213 */                     return new Relationship(_this.db, data, null, _this);
                              }
                            ;
                            }));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype__getRelationships__7() {
                          if (error) {
/*   217 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   221 */     Node.prototype.getRelationships = function Node_prototype_getRelationships__8(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_getRelationships__8",
                    line: 221
                  };
                  return __func(_, this, arguments, Node_prototype_getRelationships__8, 1, __frame, function __$Node_prototype_getRelationships__8() {
/*   222 */         return __this.all(type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   225 */     Node.prototype.outgoing = function Node_prototype_outgoing__9(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_outgoing__9",
                    line: 225
                  };
                  return __func(_, this, arguments, Node_prototype_outgoing__9, 1, __frame, function __$Node_prototype_outgoing__9() {
/*   226 */         return __this._getRelationships("outgoing", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   229 */     Node.prototype.incoming = function Node_prototype_incoming__10(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_incoming__10",
                    line: 229
                  };
                  return __func(_, this, arguments, Node_prototype_incoming__10, 1, __frame, function __$Node_prototype_incoming__10() {
/*   230 */         return __this._getRelationships("incoming", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   233 */     Node.prototype.all = function Node_prototype_all__11(type, _) {
                  var __this = this;
                  var __frame = {
                    name: "Node_prototype_all__11",
                    line: 233
                  };
                  return __func(_, this, arguments, Node_prototype_all__11, 1, __frame, function __$Node_prototype_all__11() {
/*   234 */         return __this._getRelationships("all", type, __cb(_, __frame, 1, 13, _, true));
                  });
                };
/*   237 */     Node.prototype.getRelationshipNodes = function Node_prototype_getRelationshipNodes__12(rels, _) {
                  var resp, traverseURL, _ref, _ref1, _ref2, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype_getRelationshipNodes__12",
                    line: 237
                  };
                  return __func(_, this, arguments, Node_prototype_getRelationshipNodes__12, 1, __frame, function __$Node_prototype_getRelationshipNodes__12() {
                    _this = __this;
/*   240 */         rels = ((rels instanceof Array) ? rels : [rels,]);
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_getRelationshipNodes__12() {
/*   242 */               traverseURL = (((_ref = __this._data["traverse"]) != null) ? _ref.replace("{returnType}", "node") : void 0);
/*   243 */               if (!traverseURL) {
/*   244 */                 return _(new Error("Traverse not available."));
                          }
                        ;
/*   246 */               return __this._request.post({
/*   247 */                 url: traverseURL,
/*   248 */                 json: {
/*   249 */                   max_depth: 1,
/*   250 */                   relationships: rels.map(function(rel) {
/*   251 */                     if ((typeof rel === "string")) {
/*   252 */                       return {
/*   253 */                         type: rel
                                  };
                                }
/*   255 */                      else {
/*   256 */                       return rel;
                                }
                              ;
                              })
                            }
                          }, __cb(_, __frame, 9, 15, function ___(__0, __1) {
/*   246 */                 resp = __1;
/*   261 */                 if ((resp.statusCode === 404)) {
/*   262 */                   return _(new Error((((((_ref1 = resp.body) != null) ? _ref1.message : void 0)) || "Node not found.")));
                            }
                          ;
/*   264 */                 if ((resp.statusCode !== 200)) {
/*   265 */                   return _(new Error((((((_ref2 = resp.body) != null) ? _ref2.message : void 0)) || (("Unrecognized response code: " + resp.statusCode)))));
                            }
                          ;
/*   267 */                 return _(null, resp.body.map(function(data) {
/*   268 */                   return new Node(_this.db, data);
                            }));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_getRelationshipNodes__12() {
                          if (error) {
/*   271 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   275 */     Node.prototype.path = function Node_prototype_path__13(to, type, direction, maxDepth, algorithm, _) {
                  var data, end, length, nodes, pathURL, relationships, res, start, _this, __this = this;
                  var __frame = {
                    name: "Node_prototype_path__13",
                    line: 275
                  };
                  return __func(_, this, arguments, Node_prototype_path__13, 5, __frame, function __$Node_prototype_path__13() {
                    _this = __this;
/*   278 */         if ((maxDepth == null)) {
/*   279 */           maxDepth = 1;
                    }
                  ;
/*   281 */         if ((algorithm == null)) {
/*   282 */           algorithm = "shortestPath";
                    }
                  ;
                    return (function ___(__then) {
                      (function ___(_) {
                        __tryCatch(_, function __$Node_prototype_path__13() {
/*   285 */               pathURL = (("" + __this.self) + "/path");
/*   286 */               data = {
/*   287 */                 to: to.self,
/*   288 */                 relationships: {
/*   289 */                   type: type,
/*   290 */                   direction: direction
                            },
/*   292 */                 max_depth: maxDepth,
/*   293 */                 algorithm: algorithm
                          };
/*   295 */               return __this._request.post({
/*   296 */                 url: pathURL,
/*   297 */                 json: data
                          }, __cb(_, __frame, 20, 14, function ___(__0, __1) {
/*   295 */                 res = __1;
/*   299 */                 if ((res.statusCode === status.NOT_FOUND)) {
/*   300 */                   return _(null, null);
                            }
                          ;
/*   302 */                 if ((res.statusCode !== status.OK)) {
/*   303 */                   return _(new Error(("Unrecognized response code: " + res.statusCode)));
                            }
                          ;
/*   305 */                 data = res.body;
/*   306 */                 start = new Node(__this, {
/*   307 */                   self: data.start
                            });
/*   309 */                 end = new Node(__this, {
/*   310 */                   self: data.end
                            });
/*   312 */                 length = data.length;
/*   313 */                 nodes = data.nodes.map(function(url) {
/*   314 */                   return new Node(_this, {
/*   315 */                     self: url
                              });
                            });
/*   318 */                 relationships = data.relationships.map(function(url) {
/*   319 */                   return new Relationship(_this, {
/*   320 */                     self: url,
/*   321 */                     type: type
                              });
                            });
/*   324 */                 return _(null, new Path(start, end, length, nodes, relationships));
                          }, true));
                        });
                      })(function ___(error, __result) {
                        __tryCatch(_, function __$Node_prototype_path__13() {
                          if (error) {
/*   326 */                 return _(adjustError(error));
                          }
                           else {
                            _(null, __result);
                          }
                        ;
                        });
                      });
                    })(function ___() {
                      __tryCatch(_, _);
                    });
                  });
                };
/*   330 */     return Node;
/*   332 */   })(PropertyContainer);
/*   334 */ }).call(this);
